File: src/netpbm/ppm.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <ctype.h>
#include <arpa/inet.h>

#include "netpbm/netpbm.h"
#include "netpbm/ppm.h"

void print_netppm(NETPBM *p)
{
    uint8_t *data = (uint8_t *) p->data;
    for (size_t i = 0; i < p->height; i++)
    {
        for (size_t j = 0; j < p->width * 3; j += 3)
        {
            size_t r = 0;
            size_t g = 0;
            size_t b = 0;

            if (p->bit_depth == 16)
            {
                // clamping to 0-255, but still supporting 0-65535
                r = ntohs(((uint16_t *) data)[i * 3*p->width + j + 0]) * 65535 / p->max_value / 255;
                g = ntohs(((uint16_t *) data)[i * 3*p->width + j + 1]) * 65535 / p->max_value / 255;
                b = ntohs(((uint16_t *) data)[i * 3*p->width + j + 2]) * 65535 / p->max_value / 255;
            }
            else
            {
                r = (size_t) data[i * 3*p->width + j + 0] * 255 / p->max_value;
                g = (size_t) data[i * 3*p->width + j + 1] * 255 / p->max_value;
                b = (size_t) data[i * 3*p->width + j + 2] * 255 / p->max_value;
            }

            printf("\033[48;2;%lu;%lu;%lum  \033[m", r, g, b);
        }
        puts("");
    }
}

File: src/parser/netpbm/ppm.rs
use std::io::Write;

use anyhow::Result;

use crate::{color::Color, pxbm_write, pxbm_writeln};

use super::Netpbm;

impl Netpbm {
    pub fn print_netppm(&self, writer: &mut dyn Write) -> Result<()> {
        for i in 0..self.header.height {
            for j in (0..self.header.width * 3).step_by(3) {
                let (r, g, b): (u32, u32, u32);

                if self.header.bit_depth == 16 {
                    // clamping to 0-255, but still supporting 0-65535
                    r = self.data[(i * 3 * self.header.width + j) as usize] * 65535
                        / self.header.max_value
                        / 255;
                    g = self.data[(i * 3 * self.header.width + j + 1) as usize] * 65535
                        / self.header.max_value
                        / 255;
                    b = self.data[(i * 3 * self.header.width + j + 2) as usize] * 65535
                        / self.header.max_value
                        / 255;
                } else {
                    r = self.data[(i * 3 * self.header.width + j) as usize] * 255 / self.header.max_value;
                    g = self.data[(i * 3 * self.header.width + j + 1) as usize] * 255 / self.header.max_value;
                    b = self.data[(i * 3 * self.header.width + j + 2) as usize] * 255 / self.header.max_value;
                }

                pxbm_write!(writer, "{}", Color::new(Some((r as u8, g as u8, b as u8))))?;
            }
            pxbm_writeln!(writer)?;
        }
        Ok(())
    }
}